go/arrow/compare.go

```diff
@@ -0,0 +1,58 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package arrow
+
+import (
+	"reflect"
+)
+
+// TypeEquals checks if two DataType are the same, optionally checking metadata
+// equality for STRUCT types.
+func TypeEquals(left DataType, right DataType, checkMetadata bool) bool {
```

Comment:
I'd be tempted to use (abuse?) the "functional option" pattern instead of this extra boolean:
```go
type typeEqConfig struct {
    metadata bool
}

type TypeEqualsOption func(*typeEqConfig)

func CheckMetadata() TypeEqualsOption {
    return func(cfg *typeEqConfig)
}

func TypeEquals(left, right DataType, opts ...TypeEqualsOption) bool {
    var cfg typeEqConfig
    for _, o := range opts {
        o(&cfg)
    }

    // ...
}
```

this would allow the following user code:
```go
ok1 := arrow.TypeEquals(type1, type2)
ok2 := arrow.TypeEquals(type1, type2, arrow.CheckMetaData())
```

I am not completely sold on this either.

WDYT?